Add utab monitor

azzar1 opened the pull request:
Export the info stored in /run/mounts/utab using the org.freedesktop.UDisks2.Block.Configuration property. For the moment the code to get the mount user options (x-...) has not been moved to libblockdev. I suggest to do that in a later branch as at this point other things need to moved.

vpodzime commented on src/udisksutabmonitor.c:
```diff
@@ -0,0 +1,237 @@
+  struct libmount_monitor *mn;
+
+  gboolean have_data;
+  GList *utab_entries;
```
This should be a GSList

vpodzime commented on src/udisksutabmonitor.c:
```diff
@@ -0,0 +1,237 @@
+  udisks_utab_monitor_ensure (monitor);
+
+  ret = g_list_copy (monitor->utab_entries);
+  g_list_foreach (ret, (GFunc) g_object_ref, NULL);
```
There's also ``g_list_copy_deep()`` for exactly this. And also ``g_slist_copy_deep()``.

vpodzime commented on src/udisksutabmonitor.c:
```diff
@@ -0,0 +1,237 @@
+  monitor->have_data = FALSE;
+
+  g_list_foreach (monitor->utab_entries, (GFunc) g_object_unref, NULL);
+  g_list_free (monitor->utab_entries);
```
``g_list_free_full()``/``g_slist_free_full()`` can be used here instead.

vpodzime commented on src/udisksutabmonitor.c:
```diff
@@ -0,0 +1,237 @@
+}
+
+static void
+udisks_utab_monitor_invalidate (UDisksUtabMonitor *monitor)
```
I believe there has to be some mutual exclusion mechanism involved here to prevent multiple threads from stepping on each others' toes when working with the same monitor instance (e.g. one using the data, the other doing the updates).

vpodzime commented on src/udisksutabmonitor.h:
```diff
@@ -0,0 +1,38 @@
+
+G_END_DECLS
+
+#endif /* __UDISKS_UTAB_MONITOR_H__ */
```
Missing newline here.

vpodzime commented on src/udiskslinuxblock.c:
```diff
@@ -731,6 +775,13 @@ calculate_configuration (UDisksLinuxBlock  *block,
+  for (l = entries; l != NULL; l = l->next)
+    add_utab_entry (&builder, UDISKS_UTAB_ENTRY (l->data));
+  g_list_foreach (entries, (GFunc) g_object_unref, NULL);
+  g_list_free (entries);
```
``g_(s)list_free_full()`` and same below plus possibly other places.

vpodzime commented on src/udiskslinuxblock.c:
```diff
@@ -851,6 +928,14 @@ find_configurations (gchar         *needle,
   g_list_free (entries);
 
+  /* Then the /run/mount/entries entries */
+  /* First the /etc/fstab entries */
```
copy-pasted comment that doesn't belong here?

vpodzime commented on src/udisksutabmonitor.c:
```diff
@@ -149,6 +270,7 @@ udisks_utab_monitor_constructed (GObject *object)
 
   monitor->mn = mnt_new_monitor();
+  mnt_monitor_enable_kernel (monitor->mn, TRUE);
   mnt_monitor_enable_userspace (monitor->mn, TRUE, NULL);
```
I think these two lines are worth comments explaining what is going on.

vpodzime commented on src/udiskslinuxblock.c:
```diff
@@ -933,8 +931,7 @@ find_configurations (gchar         *needle,
     add_utab_entry (&builder, UDISKS_UTAB_ENTRY (l->data));
-  g_list_foreach (entries, (GFunc) g_object_unref, NULL);
-  g_list_free (entries);
+  g_list_free_full (entries, g_object_unref);
```
Oh, I see it here now. Please squash these changes in the respective commits.

vpodzime commented on src/udisksutabentry.h:
```diff
@@ -35,4 +35,4 @@ const gchar *udisks_utab_entry_get_opts     (UDisksUtabEntry *entry);
 
-#endif /* __UDISKS_UTAB_ENTRY_H__ */
\ No newline at end of file
+#endif /* __UDISKS_UTAB_ENTRY_H__ */
```
Again, please squash the changes into the respective commits.

vpodzime commented on src/udiskslinuxblock.c:
```diff
@@ -927,7 +927,6 @@ find_configurations (gchar         *needle,
   g_list_free (entries);
 
   /* Then the /run/mount/entries entries */
-  /* First the /etc/fstab entries */
```
Same here.

vojtechtrefny commented:
I think the `Configuration` property is not a very good place for this. This more like a "state" than a "configuration". We have methods to change the configuration (`Add/Change/RemoveConfigurationItem`) -- these write the changes to `/etc/fstab` or `/etc/crypttab` -- and it will be very confusing to have a "configuration" item that is "read only". I think it would be better to create a new special property for this.

azzar1 commented:
I updated MP:
1) I squashed some commits as suggested
2) Added another dbus property named "UserspaceMountOptions" of type "as"
3) Added a read-write lock login in the monitor (even if not really needed as there is only one thread working with this monitor)
4) Updated documentation

vojtechtrefny commented:
Jenkins, ok to test.

vojtechtrefny commented on src/udisksutabmonitor.c:
```diff
@@ -0,0 +1,372 @@
+  GSource *utab_watch_source;
+
+  struct libmount_monitor *mn;
+  struct libmount_table *current_tb;
```
This should be `libmnt_monitor` and `libmnt_table`.

vojtechtrefny commented:
Thank you for changing the property. I think you could still squash most of the commits to one -- at least the *Use GSList inside UDisksUtabMonitor* and *Use a different dbus property to export the userspace mount options* shouldn't be in the final version.

Please add some tests for this too. Take a look at `dbus_tests/test_80_filesystem.py` -- simple test mounting a device with a user option and checking that the property changed should be enough.

Also please add the new dependency to `packaging/udisks2.spec` (`BuildRequires: libmount-devel` and `Requires: libmount`).

Also the libmount /util-linux 2.30 is quite new and we need to support older distributions. We should probably add an configure option to allow compiling udisks without utab monitor.

azzar1 commented:
Ok I squashed few more commits, added test (I had to allow x- mount options to be accepted by the dbus api) and made the utab monitor compiles iff libmount >= 2.30

azzar1 commented:
The only problem is that I don't know how to make the test optional. The test will fail if libmount < 2.30

azzar1 commented:
Any news about this?

vojtechtrefny commented:
Jenkins, test this please.

vojtechtrefny commented:
> The only problem is that I don't know how to make the test optional. The test will fail if libmount < 2.30

`mount --version` prints version of libmount, so you can just skip the tests if the version is too low -- we have similar check in [btrfs tests](https://github.com/storaged-project/udisks/blob/master/src/tests/dbus-tests/test_btrfs.py#L63).

> Any news about this?

There were some weird issues with the tests (some loop tests were hanging, so something that looks unrelated to the changes), I wasn't able to reproduce it manually, so it is possible that it was a problem on our machines. I will run the tests multiple times so we can be sure there isn't some hidden issue.

azzar1 commented:
Added a check for the libmount version.

vojtechtrefny commented:
Unfortunately the problem still exists -- for some reason all combinations except for CentOS 7 64bit are broken. The `test_loop_ro` from `integration-test` just hangs forever (I had to kill the tests manually, that is why it shows them as pending). It looks like completely unrelated to this change but it happened too many times (and I've never seen this issue before) to be a coincidence.
I'll try to run the test suite manually and I'll try to create some simpler reproducer. This is probably caused by some race condition -- our test machines are quite slow and because of that we see a lot more of race conditions than with manual testing.

azzar1 commented:
After some debugging a found the cause of the deadlock. Please note that g_task_get_context returns the value of g_main_context_get_thread_default at the time the GTask was created.

vpodzime commented on src/udisksthreadedjob.c:
```diff
@@ -205,7 +205,7 @@ run_task_job (GTask            *task,
     }
 
-  g_main_context_invoke (g_main_context_get_thread_default (), job_complete, job);
+  g_main_context_invoke (g_task_get_context (task), job_complete, job);
```
Thanks for spending the effort to find this nasty glitch @azzar1!

azzar1 commented:
There are still some problems with test_ext4. Do you know if this was an existing problem?

vojtechtrefny commented:
> There are still some problems with test_ext4. Do you know if this was an existing problem?

Yes, both the `No such interface 'org.freedesktop.UDisks2.Filesystem'` and `'NoneType' object has no attribute 'call_mount_sync'` are not related to your changes -- these are randomly failing for quite a long time.

vojtechtrefny commented:
Jenkins, test this please.

azzar1 commented:
Any news? Is there something else I can do in order to

vojtechtrefny commented:
Thank you again for the contribution and sorry for delays.
